Converting hardswish to strucutred kernels with metatensor support

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66899
=

Differential Revision: [D32535014](https://our.internmc.facebook.com/intern/diff/D32535014)

Gamrix committed:
Converting hardswish to strucutred kernels with metatensor support

[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Krovatkin reviewed (commented):
a few qs.

Krovatkin reviewed (approved):
:shipit:

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix commented on aten/src/ATen/native/xnnpack/Activation.cpp:
```diff
@@ -1,97 +1,94 @@
-  if (input.data_ptr() == padded_input.data_ptr()) {
-    hardswish_impl(input, input);
-    return input;
+  if (result.data_ptr() == padded_input.data_ptr()) {
```
Agreed that we shouldn't need a  functional and the inplace variant with this change.  

However, I guess I should instead be padding the result tensor and checking it against itself instead, and yes, therefore also covering the case where the out tensor is pre-padded, but not equal to the input tensor.

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix commented on aten/src/ATen/native/xnnpack/Activation.cpp:
```diff
@@ -1,97 +1,77 @@
-    input, input.suggest_memory_format());
+      input, input.suggest_memory_format());
 
-  // Don't need to allocate output if input is contiguous & already padded
-  if (input.data_ptr() == padded_input.data_ptr()) {
-    hardswish_impl(input, input);
-    return input;
+  // Don't need to allocate output if result is contiguous & already padded
+  if (mobile::is_padded_contiguous(result, result.suggest_memory_format())) {
+    hardswish_impl(padded_input, result);
```
@albanD @bdhirsh @Krovatkin Ok, rewrote this given that it seems that we only need to allocate a temporary output if the result is not properly allocated. Would still love eyes on this, because I don't know this part of the codebase well.

bdhirsh replied:
oh okay - so, I think the meta function needs to change a little. Looking at the original code for `hardswish()` (out-of-place), the logic for allocating the output looks something like:
```
  #if defined(C10_MOBILE) && defined(USE_XNNPACK)
  if (xnnpack::use_hardswish(...) && maybe_get_output().defined()) {
    // use special xnnpack logic to determine output's size and dtype
  }
  #endif
  // use TensorIteratorBase to allocate the output
```

It's a little hairy, since in the mobile case the op only decides to use TensorIterator to allocate the output some of the time.

Also, it looks like the mobile logic only happens in the `hardswish` and `hardswish_` case (and not for `out=`). So you want to check for that in the meta function. You can probably do that in the meta function with something like:
```
if (maybe_get_output().defined() && maybe_get_output().data_ptr() != input().data_ptr())` // out= case
```
I guess that would technically result in different behavior if someone tried to call `hardwish(x, out=x)`, but I'm pretty sure that pattern is buggy in a bunch of other ways throughout pytorch anyway.

Gamrix replied:
Good point. However, I thought the XNNPack version of the code should result in the same shape/size, though with the XNNPack padded memory format. I am wondering if the XNNPack implementation would deviate from the other implementation in other ways that metatensors cares about. Regardless, this deserves at least a comment in HardSwish's metafunction.

And if this is incorrect then I think the `result._copy(output)` is the thing we should scrutinize, and fix.

Gamrix replied:
I did intend on changing the logic of the function to use `xnn` if possible for the `out` case because it would make the logic cleaner, and I assumed that it was not implemented just because the person adding XNNPack support in the first place just didn't want to support it.

bdhirsh replied:
>  I am wondering if the XNNPack implementation would deviate from the other implementation in other ways that metatensors cares about. ... And if this is incorrect then I think the xnn implementation is the thing we should scrutinize, and fix first.

Yeah, I think you're right (and the code that creates the output tensor for mobile uses the same size/dtype, but just uses the custom mobile allocator). So yep, probably ignore what I said about needing mobile-specific logic in the meta function

I think the only thing that I'm worried about is that we want the above check (`mobile::is_padded_contiguous(...)`) to return true in the out-of-place `hardswish()` case. That'll only be true if the mobile allocator is used to create the output tensor (during the call to to `build_unary_op()` in the meta function). Which... I think it is? Because we only ever enter this code path when `C10_MOBILE` is set (and [there's this logic](https://github.com/pytorch/pytorch/blob/e80cb08cc80c2001f8f391fdc577521fb212da61/c10/core/CPUAllocator.cpp#L252) that sets the mobile allocator during mobile builds)

> I did intend on changing the logic of the function to use xnn if possible for the out case because it would make the logic cleaner

Ok yeah, this sounds reasonable.

Gamrix replied:
Sounds like we need a few unittests to check that the behavior of this op is as expected. For example we should that the out of place `hardswish()` doesn't do unexpected copies and goes down the path intended. I have no familiarity with how to test for that. @bdhirsh , do you have pointers for where to add this test and how to test this?

bdhirsh replied:
Not that I know of. Hey @dhruvbird, do you happen to know if we have internal tests for ops specifically on mobile? E.g. in this case, that `hardswish` goes down the XNNPACK fastpath when built with mobile (and that it continues to after this change).

I mean, I'm pretty confident that it is from staring at the code :). But an automated test would definitely help me feel better too.

dhruvbird replied:
cc @kimishpatel for tests related to XNNPACK.

@bdhirsh  if you import this PR to internal and put up a diff, then it should run tests with a bunch (like 150 models) of models that use XNNPACK (probably 50 may use XNNPACK) - if they don't crash at least you'll have some signal.

kimishpatel replied:
What @dhruvbird said should work. I wonder if we can have "code coverage" for those test to see if changed lines are covered. Not sure if there are any tools for that though.

bdhirsh commented on aten/src/ATen/native/xnnpack/Activation.cpp:
```diff
@@ -1,97 +1,77 @@
+bool use_hardswish(const Tensor& input) {
+  return xnnpack::internal::available() && (1 <= input.ndimension()) &&
+      (input.device().is_cpu()) && (kFloat == input.scalar_type()) &&
+      !input.requires_grad() && true;
```
I know it was copy-pasted, but we don't need the `&& true`, right :)

Gamrix replied:
True :P

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"




[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175963).

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"


Differential Revision: [D32175963](https://our.internmc.facebook.com/intern/diff/D32175963)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175963).

kimishpatel commented on aten/src/ATen/native/xnnpack/Activation.cpp:
```diff
@@ -1,97 +1,76 @@
-}
-
-Tensor& hardswish_impl(Tensor& input, Tensor& output) {
+void hardswish_impl(const Tensor& input, const Tensor& output) {
```
Why `const` qualifier on the output?

kimishpatel commented on aten/src/ATen/native/xnnpack/Activation.cpp:
```diff
@@ -1,97 +1,76 @@
 }
 
-Tensor& hardswish_(Tensor& input) {
+const Tensor& hardswish_out(const Tensor& input, const Tensor& result) {
```
Why is result tensor `const`?

kimishpatel commented on aten/src/ATen/native/xnnpack/Engine.h:
```diff
@@ -90,8 +90,7 @@ Tensor channel_shuffle(
 bool use_hardswish(const Tensor& input);
-Tensor hardswish(const Tensor& input);
-Tensor& hardswish_(Tensor& input);
+const Tensor& hardswish_out(const Tensor& input, const Tensor& result);
```
Returns `const Tensor&`? Is this for preserving structured kernels semantics?

kimishpatel reviewed (commented):
Left some comments on xnnpack part.

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"


Differential Revision: [D32175963](https://our.internmc.facebook.com/intern/diff/D32175963)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175963).

bdhirsh commented on aten/src/ATen/native/xnnpack/Engine.h:
```diff
@@ -90,8 +90,7 @@ Tensor channel_shuffle(
 //
 bool use_hardswish(const Tensor& input);
-Tensor hardswish(const Tensor& input);
-Tensor& hardswish_(Tensor& input);
```
Looks like there are some mobile tests at `vulkan_api_test.cpp` that expect to use this. Maybe it's worth refactoring them to use the out= variant? (Or worst case, adding this back)

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"


Differential Revision: [D32175963](https://our.internmc.facebook.com/intern/diff/D32175963)

[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"


Differential Revision: [D32175963](https://our.internmc.facebook.com/intern/diff/D32175963)

[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"


Differential Revision: [D32175963](https://our.internmc.facebook.com/intern/diff/D32175963)

[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"

=

[ghstack-poisoned]

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"

=

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32535014).

Gamrix committed:
Update on "Converting hardswish to strucutred kernels with metatensor support"

=

Differential Revision: [D32535014](https://our.internmc.facebook.com/intern/diff/D32535014)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32535014).

ezyang commented:
rip this PR
